### PR TITLE
refactor: modularize `Parse_rule`

### PR DESCRIPTION
**Pull request by brandonspark**
## What:
This PR separates the common parsing logic and formula parsing logic from `Parse_rule`, and pulls it into `Parsing_common` and `Parse_formula`.

## Why:
`Parse_rule` is huge, and I have been procrastinating cleaning it up because it's such a behemoth. Splitting it apart into two parts makes it a little less intimidating to change, and also easier to read in general.

## How:
Just pulled out the various parsing logics into the right place. Minimal actual simplfication was done.

## Test plan:
`make` 

PR checklist:

- [X] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [X] Tests included or PR comment includes a reproducible test plan
- [X] Documentation is up-to-date
- [] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [X] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

**Review by aryx — APPROVED** [review]
Good idea!
Please add a Parsing_common.mli too
(with possibly with some useful comments)

File: src/parsing/Parsing_common.ml [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+        "Pattern contains items that cannot be used together with partial \
+         matching."
+    | BadPattern (msg, pos) -> spf "%s at position %d" msg pos
+    | BadUTF8 -> "UTF8 string being matched is invalid"
+    | BadUTF8Offset ->
+        "Gets raised when a UTF8 string being matched with offset is invalid."
+    | MatchLimit ->
+        "Maximum allowed number of match attempts with\n\
+        \                      backtracking or recursion is reached during \
+         matching."
```
Comment (kopecs) [inline src/parsing/Parsing_common.ml:95]: Extra `\` here?

**Review by kopecs — APPROVED** [review]

File: src/parsing/Parsing_common.ml (hunk 1 above)
Comment (brandonspark) [reply src/parsing/Parsing_common.ml:95]: should be copied verbatim from the source file

Comment (kopecs) [reply src/parsing/Parsing_common.ml:95]: Hmm. It is, just seems off.

**Comment by aryx** [comment]
superseded by https://github.com/semgrep/semgrep/pull/9163

**Pull request closed without merging**